crates/accelerate/src/circuit_library/blocks.rs

```diff
@@ -0,0 +1,167 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::{
+    prelude::*,
+    types::{PyList, PyTuple},
+};
+use qiskit_circuit::{
+    circuit_instruction::OperationFromPython,
+    operations::{Operation, Param, StandardGate},
+    packed_instruction::PackedOperation,
+};
+use smallvec::SmallVec;
+
+use crate::{circuit_library::entanglement::get_entanglement, QiskitError};
+
+#[derive(Debug, Clone)]
+pub enum BlockOperation {
+    Standard { gate: StandardGate },
+    PyCustom { builder: Py<PyAny> },
+}
+
+impl BlockOperation {
+    pub fn assign_parameters(
+        &self,
+        py: Python,
+        params: &[&Param],
+    ) -> PyResult<(PackedOperation, SmallVec<[Param; 3]>)> {
+        match self {
+            Self::Standard { gate } => Ok((
+                (*gate).into(),
+                SmallVec::from_iter(params.iter().map(|&p| p.clone())),
```

Comment:
Maybe?:
```suggestion
                SmallVec::from_iter(params.iter().cloned()),
```

Reply:
But doesn't `p.clone()` also return a `Param` instance too? It's not a big deal, I kind of wrote this assuming clippy would be grumpy about it.